server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingTombstonesAction.java

```diff
@@ -0,0 +1,178 @@
+                    .map(allNodes::get).filter(DiscoveryNode::isMasterNode).collect(Collectors.toSet());
+
+                if (resolvedNodes.isEmpty()) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " matched no master-eligible nodes");
+                }
+
+                resolvedNodes.removeIf(n -> currentState.getVotingTombstones().contains(n));
+                if (resolvedNodes.isEmpty()) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " matched no master-eligible nodes that do not already have tombstones");
+                }
+
+                final int oldTombstoneCount = currentState.getVotingTombstones().size();
+                final int newTombstoneCount = resolvedNodes.size();
+                final int maxTombstoneCount = MAXIMUM_VOTING_TOMBSTONES_SETTING.get(currentState.metaData().settings());
+                if (oldTombstoneCount + newTombstoneCount > maxTombstoneCount) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " would add [" + newTombstoneCount + "] voting tombstones to the existing [" + oldTombstoneCount
+                        + "] which would exceed the maximum of [" + maxTombstoneCount + "] set by ["
+                        + MAXIMUM_VOTING_TOMBSTONES_SETTING.getKey() + "]");
+                }
+
+                final Builder builder = ClusterState.builder(currentState);
+                resolvedNodes.forEach(builder::addVotingTombstone);
+                final ClusterState newState = builder.build();
+                assert newState.getVotingTombstones().size() <= maxTombstoneCount;
+                return newState;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+
+                final Set<String> resolvedNodeIds = resolvedNodes.stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+
+                final Predicate<ClusterState> allNodesRemoved = new Predicate<ClusterState>() {
+                    @Override
+                    public boolean test(ClusterState clusterState) {
+                        final Set<String> votingNodeIds = clusterState.getLastCommittedConfiguration().getNodeIds();
+                        return resolvedNodeIds.stream().anyMatch(votingNodeIds::contains) == false;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "withdrawal of votes from " + resolvedNodes;
```

Comment:
why does this one need a toString()? Would be nicer to write as lambda

Reply:
Added in a debugging session, but now removed.